reframe/core/systems.py

```diff
@@ -150,17 +122,31 @@ class DeviceType(jsonext.JSONSerializable):
 
     '''
 
+    __slots__ = ('__info', 'type', 'arch')
+
+    def __deepcopy__(self, memo):
+        # This is a read-only object; simply return ourself
+        return self
+
     def __init__(self, device_info):
-        self._type = None
-        self._arch = None
-        self._num_devices = 1
-        self._info = device_info
+        self.__info = device_info
 
-        if not device_info:
-            return
+    def __getattr__(self, name):
+        if name in self.__slots__:
+            return self.__info.get(name, None)
+        else:
+            raise AttributeError(
+                f'{type(self).__qualname__!r} object has no attribute {name!r}'
+            )
 
-        for key, val in device_info.items():
-            setattr(self, f'_{key}', val)
+    def __setattr__(self, name, value):
+        if name in self.__slots__:
+            raise AttributeError(
+                f'attribute {name!r} is not writeable'
+            )
+        else:
+            # Let Python data model raise the AttributeError here
+            super().__setattr__(name, value)
```

Comment:
I think that we can factor out all these to a `_ReadOnlyInfo` object that both `ProcessorType` and `DeviceType` will inherit from. I will also rename these final objects to `ProcessorInfo` and `DeviceInfo`. "Type" can be deceiving, since it implies a `type` object.